### PR TITLE
Make masked preserve param structure

### LINKED ISSUE #271: Nesting 'masked' wrappers gives a confusing error
The _masked_ wrapper requires the model parameters to be input as a pytree. It flattens that pytree before calling init on its inner optimizer. So if you wrap masked around a optimizer that also uses masked, like lamb, optax throws a mismatched type error. It expects to receive a pytree, but instead receives a flat list of parameters. In the short term, I recommend we throw a useful error when this occurs, since it's relatively easy to mitigate this through the use of chain. In the long term I recommend rebuilding the tree in masked before passing it to the inner optimizer.

### PR DESCRIPTION
**Pull request by n2cholas**
Closes #271. 

I added a nested test that fails with the old implementation. I did not add a separate callable test since there already is one.

### CODE REVIEW

**Comment by mkunesch** [comment]
Hi @n2cholas!

Thanks a lot for this PR! I have assigned myself and will review it by Monday.

Also, just to let you know: we have highlighted your contributions on the recently updated optax contributors page
https://optax.readthedocs.io/en/latest/contributors.html.

Thank you very much for your countless contributions to optax: all the feature implementations, suggestions, bug fixes, answers to questions, ...

**Comment by n2cholas** [comment]
Sounds good @mkunesch! And thanks for the shoutout--I'm always happy to help😊!

**Review by mkunesch — CHANGES_REQUESTED** [review]
Thank you so much for this PR! I think this is a really neat implementation.

I've only really got one comment; the others are Nits/opinions.

File: optax/_src/wrappers_test.py [hunk 1]
```diff
@@ -394,6 +394,32 @@ def test_mask_fn(self):
     np.testing.assert_allclose(updates['b'][1],
                                grads['b'][1] + 0.1*params['b'][1])
 
+  @chex.all_variants
+  @parameterized.named_parameters(
+      ('scale', lambda: transform.scale(-1.)),  # stateless test
+      ('sgd', _build_sgd),  # stateful test
```
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/wrappers_test.py:400]: Totally agree! I forgot that SGD no longer has state by default after #54. 

I think `_build_simple_adam` is meant to be used as a [stateful version of `_build_sgd`](https://github.com/deepmind/optax/blob/1cb88b56c87319b4d342dfeaceef29f8217cf3f6/optax/_src/wrappers_test.py#L34-L40), however, that is actually incorrect: `_build_simple_adam` is signed gradient descent, not sgd.

It happens to work properly in the [one test it is used in](https://github.com/deepmind/optax/blob/1cb88b56c87319b4d342dfeaceef29f8217cf3f6/optax/_src/wrappers_test.py#L70-L120) because the gradient values are +/- 1.

Example illustrating that it is indeed not equivalent to sgd in general:

```py
import jax
import jax.numpy as jnp
import optax

params = jnp.full(1, 2.0)
grads = jnp.full_like(params, 3.0)

sgd = optax.sgd(1.0)
sgd_state = sgd.init(params)
updates, _ = sgd.update(grads, sgd_state)
print('SGD Update:', updates)

adam = optax.adam(1.0, b1=0., b2=0.)
adam_state = adam.init(params)
updates, _ = adam.update(grads, adam_state)
print('Adam Update: ', updates)
```
Output:
```
>> SGD Update: [-3.]
>> Adam Update:  [-1.]
```

I think we should replace `_build_simple_adam` with `_build_stateful_sgd` where the difference is `momentum=0.0`.  Then, I can fix tests that use `_build_sgd` and `_build_simple_adam` as well as add the test you suggested.

How does that sound?

File: optax/_src/wrappers_test.py [hunk 2]
```diff
@@ -394,6 +394,32 @@ def test_mask_fn(self):
...
+
+    outer_mask = lambda p: jax.tree_map(lambda x: x.ndim > 1, p)
+    inner_mask = jax.tree_map(lambda _: True, params)
+    inner_mask['linear_2'] = False
+
+    inner = wrappers.masked(opt_builder(), inner_mask)
+    init_fn, update_fn = wrappers.masked(inner, outer_mask)
+
+    input_updates = jax.tree_map(jnp.ones_like, params)
+    correct_updates = jax.tree_map(lambda x: x, input_updates)
```
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/wrappers_test.py]: Done!

File: optax/_src/wrappers_test.py [hunk 3]
```diff
@@ -394,6 +394,32 @@ def test_mask_fn(self):
     np.testing.assert_allclose(updates['b'][1],
                                grads['b'][1] + 0.1*params['b'][1])
 
+  @chex.all_variants
+  @parameterized.named_parameters(
+      ('scale', lambda: transform.scale(-1.)),  # stateless test
+      ('sgd', _build_sgd),  # stateful test
+  )
+  def test_nested_mask(self, opt_builder):
```
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/wrappers_test.py:404]: Done!

File: optax/_src/wrappers_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mkunesch) [reply optax/_src/wrappers_test.py:402]: Hi! Yes, that sounds good to me! Let's do it in a separate PR so that we can roll back the changes separately if necessary. Would you like to do the PR for that or shall I? (I'm happy either way, just want to make sure we don't work on it at the same time)

**Review by mkunesch — COMMENTED** [review]
Thanks a lot - this looks great to me! I'll wait with approval until we have made the `_build_sgd` change.

File: optax/_src/wrappers_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/wrappers_test.py:402]: I'll open a PR!

[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/wrappers_test.py]: Done.

File: optax/_src/wrappers_test.py [hunk 4]
```diff
@@ -396,6 +398,44 @@ def test_mask_fn(self):
...
+    correct_updates = copy.deepcopy(input_updates)
+    correct_updates['linear_1']['w'] *= -1.0
+    correct_updates['linear_3']['w'] *= -1.0
+
+    state = self.variant(init_fn)(params)
+    updates, state = self.variant(update_fn)(input_updates, state, params)
+    chex.assert_trees_all_close(updates, correct_updates)
+
+  @chex.all_variants
+  def test_masked_state_structure(self):
```
Comment (mkunesch) [inline optax/_src/wrappers_test.py:429]: Nice, thanks a lot - it's great to have this as a separate unit test!

**Review by mkunesch — APPROVED** [review]
Looks great to me. Thanks a lot!

**Comment by mkunesch** [comment]
Thanks a lot for the PR again! Took a bit longer to merge due to internal tests but it's all done now.

I was looking at the PR with @hbq1 and he pointed out that it might be good to mention in the docstring that `mask` is expected to be static. @n2cholas do you agree? Would you like to do that in a new PR or shall I add it?
